symbolic-base/src/ZkFold/Symbolic/Compiler.hs

```diff
@@ -41,88 +37,50 @@ import           ZkFold.Symbolic.MonadCircuit               (MonadCircuit (..))
     6. ZkFold.Symbolic.Compiler
 -}
 
-forceOne :: (Symbolic c, Traversable f) => c f -> c f
-forceOne r = fromCircuitF r (\fi -> for fi $ \i -> constraint (\x -> x i - one) $> i)
-
--- | Arithmetizes an argument by feeding an appropriate amount of inputs.
-solder ::
-    forall a c p f s .
-    ( c ~ ArithmeticCircuit a p (Layout s)
-    , SymbolicData f
-    , Context f ~ c
-    , Support f ~ s
-    , SymbolicInput s
-    , Context s ~ c
-    , Symbolic c
-    ) => f -> c (Layout f)
-solder f = fromCircuit2F (pieces f input) b $ \r (Par1 i) -> do
-    constraint (\x -> one - x i)
-    return r
+-- | Arithmetizes an argument by feeding the given support.
+solderWith ::
+  ( SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Symbolic c
+  ) => c (Layout s) -> f -> c (Layout f)
+solderWith sLayout f = fromCircuit2F (pieces f input) b $ \r (Par1 i) -> do
+  constraint (\x -> one - x i)
+  return r
   where
     Bool b = isValid input
-    input = restore @(Support f) $ const idCircuit
+    input = restore (const sLayout)
 
--- | Compiles function `f` into an arithmetic circuit with all outputs equal to 1.
-compileForceOne ::
-    forall a c p f s l y .
-    ( c ~ ArithmeticCircuit a p l
-    , Arithmetic a
-    , Binary a
-    , SymbolicData f
-    , Context f ~ c
-    , Support f ~ s
-    , SymbolicInput s
-    , Context s ~ c
-    , Layout s ~ l
-    , Binary (Rep p)
-    , Binary (Rep l)
-    , Ord (Rep l)
-    , SymbolicData y
-    , Context y ~ c
-    , Support y ~ Proxy c
-    , Layout f ~ Layout y
-    , Traversable (Layout y)
-    ) => f -> y
-compileForceOne = restore . const . optimize . forceOne . solder @a
+finalRestore ::
+  ( SymbolicData y, Context y ~ c, Support y ~ Proxy c
+  , c ~ ArithmeticCircuit a p i) => c (Layout y) -> y
+finalRestore = restore . const . optimize
 
--- | Compiles function `f` into an arithmetic circuit.
-compile ::
-    forall a c p f s l y .
-    ( c ~ ArithmeticCircuit a p l
-    , SymbolicData f
-    , Context f ~ c
-    , Support f ~ s
-    , SymbolicInput s
-    , Context s ~ c
-    , Layout s ~ l
-    , SymbolicData y
-    , Context y ~ c
-    , Support y ~ Proxy c
-    , Layout f ~ Layout y
-    , Symbolic c
-    ) => f -> y
-compile = restore . const . optimize . solder @a
+compile :: forall a y f c s p.
+  ( SymbolicData f, Context f ~ c, Support f ~ s, SymbolicInput s, Context s ~ c
+  , c ~ ArithmeticCircuit a p (Layout s), Arithmetic a, Binary a, Binary (Rep p)
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c, Layout y ~ Layout f)
+  => f -> y
+compile = finalRestore . solderWith idCircuit
```

Comment:
I propose adding `compileWith` function that has parameters: a circuit for `solderWith` and a transformation after the `solderWith` function (like `forceOne` or `guessOutput`).